Automattic/wp-calypso
# composite-checkout: Add contact field validation

#### Changes proposed in this Pull Request

This PR modifies the contact form so that the fields each have their own validation and the "Continue" button is disabled until all the required fields are filled.

We need to keep three pieces of state for each field: its value, its "touched" state (true if it has been edited), and its valid state (can be anything but for most fields it's just false if the field is empty). In order to initialize all these values, we need to control the store where the state lives. To do this, this PR creates a new data store (using the data package from composite-checkout) with its own reducer, selectors, and actions for the contact form.

In order to both initialize and use the new store during the first render, this PR also splits the `WPCOMCheckout` component into two: `WPCheckoutWrapper` (the new main export of composite-checkout-wpcom) and `WPCheckout` (which contains the steps).

In order to simplify the form and the data, and because they should never both be displayed at the same time, we also merge `postalCode` and `zipCode`, as well as `state` and `province`.

#### Testing instructions

- Start calypso and visit the checkout page with the `composite-checkout-wpcom` flag.
- Fill out the contact form (step 2).
- Verify that the "Continue" button is disabled until all the fields are filled.
- Empty completed fields and verify that each one turns red when it becomes empty.
- Change the data being fed to the form so that it displays the domains fields (modify `areDomainsInLineItems` in `wp-calypso/packages/composite-checkout-wpcom/src/hooks/has-domains.js` so that it always returns true).
- Verify that the "Continue" button is disabled until all the fields are filled.
- Empty completed fields and verify that each one turns red when it becomes empty and displays a warning message.
